### PR TITLE
[#IC-238] db schema migrations for `io-subscrioption-migrations`

### PR DESCRIPTION
**Pull request by balanza**
### List of changes

see commit list

### Type of changes

- [ ] Add new resources
- [x] Update configuration to existing resources
- [ ] Remove existing resources

### Env to apply

- [ ] DEV
- [ ] UAT
- [x] PROD

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [x] No

### Does this introduce an unwanted change on infrastructure? Check terraform plan execution result

- [ ] Yes
- [ ] No

### Other information

---

### How to apply

After PR is approved
1. run deploy pipeline from Azure DevOps [io-infra.deploy](https://dev.azure.com/pagopaspa/io-iac-projects/_build?definitionId=218)
2. select PR branch
3. wait for approval

Depends on https://github.com/pagopa/azurerm/pull/179
Depends on https://github.com/pagopa/io-infra/pull/103

### CODE REVIEW

**Review by uolter — APPROVED** [review]

**Review by uolter — DISMISSED** [review]

**Review by uolter — APPROVED** [review]

**Review by uolter — APPROVED** [review]

File: .devops/code-review-pipelines.yml [hunk 1]
```diff
@@ -55,3 +55,12 @@ stages:
...
               ENVIRONMENT: prod
               WORKINGDIR: 'src/core'
+      - job: validate_migration
+        steps:
+          - checkout: self
+          - script: |
+              set -e
+              cd src/psql/selfcare/subscription-migrations
+              echo "Validating migrations"
+              sh ./flyway.sh validate PROD-IO && echo "Validation successful" || echo "Validation failed"; exit 1;
```
Comment (pasqualedevita) [inline .devops/code-review-pipelines.yml:65]:
```suggestion
              sh ./flyway.sh validate PROD-IO db && echo "Validation successful" || echo "Validation failed"; exit 1;
```
kudos @balanza for the initiative :)

File: .devops/deploy-pipelines.yml [hunk 2]
```diff
@@ -70,3 +70,16 @@ stages:
...
+        dependsOn: apply
+        environment: 'PROD'
+        strategy:
+          runOnce:
+            deploy:
+              steps:
+                - checkout: self
+                - script: |
+                    cd src/psql/selfcare/subscription-migrations
+                    bash ./flyway.sh migrate PROD-IO
```
Comment (pasqualedevita) [inline .devops/deploy-pipelines.yml:84]:
```suggestion
                    bash ./flyway.sh migrate PROD-IO db
```

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: .devops/deploy-pipelines.yml (hunk 2 above)
Comment (balanza) [reply .devops/deploy-pipelines.yml:84]: I actually hardcoded `db` into the shell script - I can move it out again

File: .devops/code-review-pipelines.yml [hunk 3]
```diff
@@ -55,3 +55,12 @@ stages:
               SUBSCRIPTION: PROD-IO-SERVICE-CONN
               ENVIRONMENT: prod
               WORKINGDIR: 'src/core'
+      - job: validate_migration
+        steps:
+          - checkout: self
+          - script: |
```
Comment (pasqualedevita) [inline .devops/code-review-pipelines.yml:61]: u need a `AzureCLI@2 `task because script needs to run az cli commands

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

**Review by uolter — DISMISSED** [review]

File: src/psql/selfcare/subscription-migrations/flyway.sh [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  WORKDIR=${WORKDIR//\\//}
+fi
+
+echo "Running Flyway docker container"
+docker run --rm --network=host -v "${WORKDIR}/migrations/${DATABASE}":/flyway/sql \
+  flyway/flyway:"${FLYWAY_DOCKER_TAG}" \
+  -url="${FLYWAY_URL}" -user="${FLYWAY_USER}" -password="${FLYWAY_PASSWORD}" \
+  -validateMigrationNaming=true \
+  -placeholders.flywayUser="${administrator_login}" \
+  -placeholders.fnsubsmigrationsUserPassword="${}" \
```
Comment (pasqualedevita) [inline src/psql/selfcare/subscription-migrations/flyway.sh:83]:
```suggestion
  -placeholders.fnsubsmigrationsUserPassword="${FNSUBSMIGRATIONS_USER_PASSWORD}" \
```

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

**Comment by balanza** [comment]
I rebased history so commit list is readable, some comment may refer to a missing commit

File: .devops/code-review-pipelines.yml [hunk 5]
```diff
@@ -55,3 +55,19 @@ stages:
...
+              azureSubscription: PROD-IO-SERVICE-CONN
+              scriptType: 'bash'
+              scriptLocation: 'inlineScript'
+              failOnStandardError: false
+              inlineScript: |
+                set -e
+                cd src/psql/selfcare/subscription-migrations
+                echo "Validating migrations"
+                ./flyway.sh validate PROD-IO db && echo "Validation successful" || { echo "Validation failed"; exit 1; }
+                echo $administrator_login
```
Comment (pasqualedevita) [inline .devops/code-review-pipelines.yml:73]:
```suggestion
```
maybe we can remove this :)

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: .devops/code-review-pipelines.yml [hunk 6]
```diff
@@ -55,3 +55,19 @@ stages:
...
+            inputs:
+              azureSubscription: PROD-IO-SERVICE-CONN
+              scriptType: 'bash'
+              scriptLocation: 'inlineScript'
+              failOnStandardError: false
+              inlineScript: |
+                set -e
+                cd src/psql/selfcare/subscription-migrations
+                echo "Validating migrations"
+                ./flyway.sh validate PROD-IO db && echo "Validation successful" || { echo "Validation failed"; exit 1; }
```
Comment (pasqualedevita) [inline .devops/code-review-pipelines.yml:72]:
```suggestion
                ./flyway.sh info PROD-IO db && echo "Validation successful" || { echo "Validation failed"; exit 1; }
```

> from flyaway docs: https://flywaydb.org/documentation/command/validate
> The validate mechanism checks if the migration locally still has the same checksum as the migration already executed in the database.
> 

So when we add sql migrations is expected that validate fails. Probably we need to add `-ignorePendingMigrations=true` or use `info` command

**Review by pasqualedevita — CHANGES_REQUESTED** [review]

File: .devops/code-review-pipelines.yml (hunk 6 above)
Comment (balanza) [reply .devops/code-review-pipelines.yml:72]: Dunno, let's try - good insight indeed!

**Review by pasqualedevita — DISMISSED** [review]

**Review by uolter — DISMISSED** [review]

**Review by pasqualedevita — DISMISSED** [review]

**Review by pasqualedevita — APPROVED** [review]

**Pull request merged**